From 8f23a2c2d14fd799813134e995c160354d75d3a0 Mon Sep 17 00:00:00 2001
From: Hans Wennborg <hans@chromium.org>
Date: Thu, 25 Aug 2022 11:44:42 +0000
Subject: [PATCH] Reland "[clang] Make update.py verify that the expected and
 real versions match"

This is a reland of commit dc9652a82d4d46a4033e25e95d68854a91c04f28
It includes a fix to package.py to not use --print-version.

Original change's description:
> [clang] Make update.py verify that the expected and real versions match
>
> To fix problems when building with a new version of update.py without
> having run 'gclient sync', which could e.g. lead to PCH files built with
> the wrong clang version being left in the tree (see bug).
>
> Bug: 608105
> Change-Id: Ia5b70506be3f6a89d17742b6b2e0bf3271585d31
> Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3847019
> Commit-Queue: Hans Wennborg <hans@chromium.org>
> Reviewed-by: Nico Weber <thakis@chromium.org>
> Cr-Commit-Position: refs/heads/main@{#1038229}

Bug: 608105
Change-Id: Ia65e4e9350a3048298b06d47968eabc641609eb8
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3854616
Reviewed-by: Nico Weber <thakis@chromium.org>
Commit-Queue: Hans Wennborg <hans@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1039192}
---
 tools/clang/scripts/package.py | 10 ++--------
 tools/clang/scripts/update.py  | 17 ++++++++++++-----
 2 files changed, 14 insertions(+), 13 deletions(-)

diff --git a/tools/clang/scripts/package.py b/tools/clang/scripts/package.py
index 380d14c44c970..d30f85dc1ce2b 100755
--- a/tools/clang/scripts/package.py
+++ b/tools/clang/scripts/package.py
@@ -19,7 +19,7 @@
 import tarfile
 import time
 
-from update import RELEASE_VERSION, STAMP_FILE
+from update import PACKAGE_VERSION, RELEASE_VERSION, STAMP_FILE
 
 # Path constants.
 THIS_DIR = os.path.dirname(__file__)
@@ -66,12 +66,6 @@ def PrintTarProgress(tarinfo):
   return tarinfo
 
 
-def GetExpectedStamp():
-  rev_cmd = [sys.executable, os.path.join(THIS_DIR, 'update.py'),
-             '--print-revision']
-  return str(subprocess.check_output(rev_cmd).decode()).rstrip()
-
-
 def GetGsutilPath():
   if not 'find_depot_tools' in sys.modules:
     sys.path.insert(0, os.path.join(CHROMIUM_DIR, 'build'))
@@ -196,7 +190,7 @@ def main():
     print('--build-mac-arm only valid on intel to cross-build arm')
     return 1
 
-  expected_stamp = GetExpectedStamp()
+  expected_stamp = PACKAGE_VERSION
   pdir = 'clang-' + expected_stamp
   print(pdir)
 
diff --git a/tools/clang/scripts/update.py b/tools/clang/scripts/update.py
index 3114812d83d9a..e6b1fdc633d41 100755
--- a/tools/clang/scripts/update.py
+++ b/tools/clang/scripts/update.py
@@ -340,6 +340,11 @@ def main():
     print(RELEASE_VERSION)
     return 0
 
+  if args.output_dir:
+    global LLVM_BUILD_DIR, STAMP_FILE
+    LLVM_BUILD_DIR = os.path.abspath(args.output_dir)
+    STAMP_FILE = os.path.join(LLVM_BUILD_DIR, 'cr_build_revision')
+
   if args.print_revision:
     if args.llvm_force_head_revision:
       force_head_revision = ReadStampFile(FORCE_HEAD_REVISION_FILE)
@@ -349,6 +354,13 @@ def main():
       print(force_head_revision)
       return 0
 
+    stamp_version = ReadStampFile(STAMP_FILE).partition(',')[0]
+    if PACKAGE_VERSION != stamp_version:
+      print('The expected clang version is %s but the actual version is %s' %
+            (PACKAGE_VERSION, stamp_version))
+      print('Did you run "gclient sync"?')
+      return 1
+
     print(PACKAGE_VERSION)
     return 0
 
@@ -356,11 +368,6 @@ def main():
     print('--llvm-force-head-revision can only be used for --print-revision')
     return 1
 
-  if args.output_dir:
-    global LLVM_BUILD_DIR, STAMP_FILE
-    LLVM_BUILD_DIR = os.path.abspath(args.output_dir)
-    STAMP_FILE = os.path.join(LLVM_BUILD_DIR, 'cr_build_revision')
-
   return UpdatePackage(args.package, args.host_os)
 
 
